packages/protocol/src/helpers/AddrMapper.sol

```diff
@@ -1,59 +1,94 @@
- * Required for getting contract addresses for some providers and flashloan providers
+ * @title AddrMapper
+ * @author Fujidao Labs
+ * @notice Contract that stores and returns addresses mappings
+ * Required for getting contract addresses for some providers and flashloan providers.
  */
-contract AddrMapper is IAddrMapper, Ownable {
-  string public mapperName;
-  // key address => returned address
-  // (e.g. public erc20 => protocol Token)
-  mapping(address => address) private _addrMapping;
-  // key1 address, key2 address => returned address
-  // (e.g. collateral erc20 => borrow erc20 => Protocol market)
-  mapping(address => mapping(address => address)) private _addrNestedMapping;
-
-  constructor(string memory _mapperName) {
-    mapperName = _mapperName;
+
+import {SystemAccessControl} from "../access/SystemAccessControl.sol";
+import {IAddrMapper} from "../interfaces/IAddrMapper.sol";
+
+contract AddrMapper is IAddrMapper, SystemAccessControl {
+  // provider name => key address => returned address
+  // (e.g. Compound_V2 => public erc20 => protocol Token)
+  mapping(string => mapping(address => address)) private _addrMapping;
+  // provider name => key1 address => key2 address => returned address
+  // (e.g. Compound_V3 => collateral erc20 => borrow erc20 => Protocol market)
+  mapping(string => mapping(address => mapping(address => address))) private _addrNestedMapping;
+
+  string[] private _providerNames;
+
+  mapping(string => bool) private _isProviderNameAdded;
+
+  constructor(address chief) SystemAccessControl(chief) {}
+
+  function getProviders() public view returns (string[] memory) {
+    return _providerNames;
   }
 
-  function getAddressMapping(address inputAddr) external view override returns (address) {
-    return _addrMapping[inputAddr];
+  function getAddressMapping(string memory providerName, address inputAddr)
+    external
+    view
+    override
+    returns (address)
+  {
+    return _addrMapping[providerName][inputAddr];
   }
 
-  function getAddressNestedMapping(address inputAddr1, address inputAddr2)
+  function getAddressNestedMapping(
+    string memory providerName,
+    address inputAddr1,
+    address inputAddr2
+  )
     external
     view
     override
     returns (address)
   {
-    return _addrNestedMapping[inputAddr1][inputAddr2];
+    return _addrNestedMapping[providerName][inputAddr1][inputAddr2];
   }
 
   /**
    * @dev Adds an address mapping.
+   * Requirements:
+   * - ProviderName should be formatted: Name_Version or Name_Version_Asset
    */
-  function setMapping(address keyAddr, address returnedAddr) public override onlyOwner {
-    _addrMapping[keyAddr] = returnedAddr;
+  function setMapping(string memory providerName, address keyAddr, address returnedAddr)
+    public
+    override
+    onlyTimeLock
+  {
+    if (!_isProviderNameAdded[providerName]) {
+      _isProviderNameAdded[providerName] = true;
+    }
+    _addrMapping[providerName][keyAddr] = returnedAddr;
```

Comment:
need to push `providerName` to `_providerNames`

Reply:
good catch this is fixed.